smart_open/gcs.py

```diff
@@ -0,0 +1,524 @@
+
+import io
+import logging
+import sys
+
+from google.cloud import storage, exceptions
+import google.auth.transport.requests as google_requests
+import six
+
+import smart_open.bytebuffer
+
+logger = logging.getLogger(__name__)
+
+READ_BINARY = 'rb'
+WRITE_BINARY = 'wb'
+MODES = (READ_BINARY, WRITE_BINARY)
+"""Allowed I/O modes for working with GCS."""
+
+_BINARY_TYPES = (six.binary_type, bytearray)
+"""Allowed binary buffer types for writing to the underlying GCS stream"""
+if sys.version_info >= (2, 7):
+    _BINARY_TYPES = (six.binary_type, bytearray, memoryview)
+
+_UNKNOWN_FILE_SIZE = '*'
+
+_RESUME_INCOMPLETE = 308
+
+BINARY_NEWLINE = b'\n'
+
+SUPPORTED_SCHEMES = ("gcs", "gs")
+
+REQUIRED_CHUNK_MULTIPLE = 256 * 1024
+"""Google requires you to upload in multiples of 256 KB, except for the last part."""
+
+MIN_MIN_PART_SIZE = 256 * 1024
+"""The absolute minimum permitted by Google."""
+DEFAULT_BUFFER_SIZE = 256 * 1024
+
+START = 0
+CURRENT = 1
+END = 2
+WHENCE_CHOICES = [START, CURRENT, END]
+
+SUCCESSFUL_STATUS_CODES = (200, 201)
+
+
+def clamp(value, minval, maxval):
+    return max(min(value, maxval), minval)
+
+
+def make_range_string(start, stop=None, end=_UNKNOWN_FILE_SIZE):
+    #
+    # https://cloud.google.com/storage/docs/xml-api/resumable-upload#step_3upload_the_file_blocks
+    #
+    if stop is None:
+        return 'bytes %d-/%s' % (start, end)
+    return 'bytes %d-%d/%s' % (start, stop, end)
+
+
+class UploadFailedError(Exception):
+    """Raised when a multi-part upload to GCS returns a failed response status code."""
+
+
+def open(
+        bucket_id,
+        blob_id,
+        mode,
+        buffer_size=DEFAULT_BUFFER_SIZE,
+        min_part_size=MIN_MIN_PART_SIZE,
+        client=None,  # type: storage.Client
+        ):
+    """Open an GCS blob for reading or writing.
+
+    Parameters
+    ----------
+    bucket_id: str
+        The name of the bucket this object resides in.
+    blob_id: str
+        The name of the blob within the bucket.
+    mode: str
+        The mode for opening the object.  Must be either "rb" or "wb".
+    buffer_size: int, optional
+        The buffer size to use when performing I/O.
+    min_part_size: int, optional
+        The minimum part size for multipart uploads.  For writing only.
+    client: object, optional
+        The GCS client to use when working with google-cloud-storage.
+
+    """
+    if mode == READ_BINARY:
+        return SeekableBufferedInputBase(
+            bucket_id,
+            blob_id,
+            buffer_size=buffer_size,
+            line_terminator=BINARY_NEWLINE,
+            client=client,
+        )
+    elif mode == WRITE_BINARY:
+        return BufferedOutputBase(
+            bucket_id,
+            blob_id,
+            min_part_size=min_part_size,
+            client=client,
+        )
+    else:
+        raise NotImplementedError('GCS support for mode %r not implemented' % mode)
+
+
+class RawReader(object):
```

Comment:
This looks like a useless class to me. You can replace it easily with:

```python
alternative = io.BytesIO(gcs_blob.download_as_string())
```

and it will behave the same way.

BTW, how large are these blobs? Is it acceptable to store them entirely in memory? I suspect the answer is no.

Reply:
Great suggestion, RawReader has been removed